### PR TITLE
Defeat compiler optimization that is removing the core of the libpng fuzzer as "dead code"

### PR DESCRIPTION
**Pull request by glennrp**

### CODE REVIEW

**Comment by glennrp** [comment]
The "coverage" reports indicate that the core of the libpng fuzzer is never executed.  I believe it is because of "dead code elimination" by the compiler, because the libpng row data is never used.

**Comment by glennrp** [comment]
By "core of the libpng fuzzer" I am referring to this loop:
```
  for (int pass = 0; pass < passes; ++pass) {
    for (png_uint_32 y = 0; y < height; ++y) {
      png_read_row(png_handler.png_ptr,
                   static_cast<png_bytep>(png_handler.row_ptr), nullptr);
    }
  }

```

**Comment by Dor1s** [comment]
Thanks for the clarification, @glennrp. I got it, and started to look into that issue using another coverage tool: https://clang.llvm.org/docs/SourceBasedCodeCoverage.html

However, that tool gave me even more confusing result. I'll give it another shot tomorrow.

**Comment by Dor1s** [comment]
It's strange indeed. I added a few debugging printf calls:

```cpp
--- a/contrib/oss-fuzz/libpng_read_fuzzer.cc
+++ b/contrib/oss-fuzz/libpng_read_fuzzer.cc
@@ -17,7 +17,7 @@
 #include <stddef.h>
 #include <stdint.h>
 #include <string.h>
-
+#include <stdio.h>
 #include <vector>
 
 #define PNG_INTERNAL
@@ -167,11 +167,14 @@ extern "C" int LLVMFuzzerTestOneInput(const uint8_t* data, size_t size) {
   }
 
   int passes = png_set_interlace_handling(png_handler.png_ptr);
+  printf("passes: %i\n", passes);
   png_start_read_image(png_handler.png_ptr);
 
   /* To do: prevent the optimizer from removing this code entirely */
   for (int pass = 0; pass < passes; ++pass) {
+    printf("height: %u\n", height);
     for (png_uint_32 y = 0; y < height; ++y) {
+      printf("  going to call png_read_row...\n");
       png_read_row(png_handler.png_ptr,
                    static_cast<png_bytep>(png_handler.row_ptr), nullptr);
     }
```

And then run the fuzzer over minimized corpus:

```shell
# ./libpng_read_fuzzer -runs=0 ./Corpus/ > trace.txt
INFO: Seed: 463930183
INFO: Loaded 1 modules   (3776 guards): 3776 [0x8f82e0, 0x8fbde0), 
INFO:     1634 files found in ./Corpus/
INFO: -max_len is not provided; libFuzzer will not generate inputs larger than 134287 bytes
INFO: seed corpus: files: 1634 min: 1b max: 134287b total: 1362023b rss: 32Mb
#1636	INITED cov: 906 ft: 4249 corp: 1633/1329Kb exec/s: 0 rss: 357Mb
#1636	DONE   cov: 906 ft: 4249 corp: 1633/1329Kb exec/s: 0 rss: 357Mb
Done 1636 runs in 0 second(s)
```


As a result, `trace.txt` contains only 52 lines all printed before the loop:

```
passes: 1
passes: 1
passes: 7
passes: 1
passes: 1
passes: 7
passes: 7
passes: 1
passes: 1
passes: 1
passes: 1
passes: 7
passes: 7
passes: 7
passes: 1
passes: 1
passes: 1
passes: 1
passes: 7
passes: 7
passes: 1
passes: 1
passes: 7
passes: 1
passes: 7
passes: 1
passes: 7
passes: 1
passes: 1
passes: 1
passes: 1
passes: 1
passes: 7
passes: 7
passes: 1
passes: 1
passes: 7
passes: 1
passes: 1
passes: 1
passes: 1
passes: 1
passes: 1
passes: 1
passes: 1
passes: 1
passes: 1
passes: 1
passes: 1
passes: 1
passes: 1
passes: 1
```

**Comment by Dor1s** [comment]
FTR, clang source based coverage is failing during report generation. Probably due to some weird optimizations as well.

**Comment by Dor1s** [comment]
Even with `-O0` I'm getting the same trace...

**Comment by kcc** [comment]
Maybe longjmp is involved here somehow?

**Comment by Dor1s** [comment]
After applying the following change:

```cpp
-  int passes = png_set_interlace_handling(png_handler.png_ptr);
+  volatile int passes = png_set_interlace_handling(png_handler.png_ptr);
+  printf("passes before png_start_read_image: %i\n", passes);
   png_start_read_image(png_handler.png_ptr);
+  printf("passes after png_start_read_image: %i\n", passes);
```

I realized that we never execute code after `png_start_read_image`. After some debugging, I noticed that we do enter `png_start_read_image`, but after that we end up with following:

```
(gdb) bt
#0  0x00000000005bf555 in png_longjmp () at pngerror.c:763
#1  0x00000000005bb27b in png_default_error () at pngerror.c:754
#2  0x00000000005bb243 in png_error () at pngerror.c:88
#3  0x0000000000519f97 in user_read_data(png_struct_def*, unsigned char*, unsigned long) ()
    at /src/libpng/contrib/oss-fuzz/libpng_read_fuzzer.cc:54
#4  0x00000000005e7a38 in png_read_data () at pngrio.c:37
#5  0x00000000006240b9 in png_crc_read () at pngrutil.c:201
#6  0x00000000006242e2 in png_crc_finish () at pngrutil.c:226
#7  0x00000000006333d7 in png_handle_eXIf () at pngrutil.c:2065
#8  0x00000000005ccbfb in png_read_info () at pngread.c:180
#9  0x000000000051b945 in LLVMFuzzerTestOneInput () at /src/libpng/contrib/oss-fuzz/libpng_read_fuzzer.cc:140
#10 0x00000000005420f1 in ExecuteCallback () at /src/libfuzzer/FuzzerLoop.cpp:463
#11 0x00000000005407c7 in RunOne () at /src/libfuzzer/FuzzerLoop.cpp:392
#12 0x0000000000544cf0 in ReadAndExecuteSeedCorpora () at /src/libfuzzer/FuzzerLoop.cpp:640
#13 0x0000000000545419 in Loop () at /src/libfuzzer/FuzzerLoop.cpp:660
#14 0x000000000052cfe7 in FuzzerDriver () at /src/libfuzzer/FuzzerDriver.cpp:736
#15 0x00000000005203b9 in main () at /src/libfuzzer/FuzzerMain.cpp:20
(gdb) x/32i $rip
=> 0x5bf555 <png_longjmp+373>:	mov    -0x48(%rbp),%rax
   0x5bf559 <png_longjmp+377>:	mov    (%rax),%rdi
   0x5bf55c <png_longjmp+380>:	mov    -0xc(%rbp),%esi
   0x5bf55f <png_longjmp+383>:	mov    -0x40(%rbp),%rcx
   0x5bf563 <png_longjmp+387>:	callq  *%rcx
   0x5bf565 <png_longjmp+389>:	callq  0x4e9f10 <__asan_handle_no_return()>
   0x5bf56a <png_longjmp+394>:	callq  0x41b790 <abort@plt>
   0x5bf56f:	nop
```

So yeah, longjmp seems to be involved.

**Comment by Dor1s** [comment]
**TL;DR** the error handler seems to jump into `PNG_CLEANUP    return 0;` code, so we silently leave `LLVMFuzzerTestOneInput`.

**Comment by Dor1s** [comment]
In case if you don't believe it, the following code is modifying the stack and jumping into there:

```
(gdb) bt
#0  __longjmp () at ../sysdeps/x86_64/__longjmp.S:29
#1  0x00007ffff6ce0289 in __libc_siglongjmp (env=0x619000002880, val=1) at longjmp.c:38
#2  0x00000000005bf565 in png_longjmp () at pngerror.c:763
#3  0x00000000005bb27b in png_default_error () at pngerror.c:754
#4  0x00000000005bb243 in png_error () at pngerror.c:88
#5  0x00000000005bd987 in png_chunk_error () at pngerror.c:490
#6  0x000000000062e7cb in png_handle_sPLT () at pngrutil.c:1674
#7  0x00000000005ccf25 in png_read_info () at pngread.c:230
#8  0x000000000051b945 in LLVMFuzzerTestOneInput () at /src/libpng/contrib/oss-fuzz/libpng_read_fuzzer.cc:140
#9  0x00000000005420f1 in ExecuteCallback () at /src/libfuzzer/FuzzerLoop.cpp:463
#10 0x00000000005407c7 in RunOne () at /src/libfuzzer/FuzzerLoop.cpp:392
#11 0x0000000000544cf0 in ReadAndExecuteSeedCorpora () at /src/libfuzzer/FuzzerLoop.cpp:640
#12 0x0000000000545419 in Loop () at /src/libfuzzer/FuzzerLoop.cpp:660
#13 0x000000000052cfe7 in FuzzerDriver () at /src/libfuzzer/FuzzerDriver.cpp:736
#14 0x00000000005203b9 in main () at /src/libfuzzer/FuzzerMain.cpp:20
(gdb) x/32i $rip
=> 0x7ffff6ce02a0 <__longjmp>:	mov    0x30(%rdi),%r8
   0x7ffff6ce02a4 <__longjmp+4>:	mov    0x8(%rdi),%r9
   0x7ffff6ce02a8 <__longjmp+8>:	mov    0x38(%rdi),%rdx
   0x7ffff6ce02ac <__longjmp+12>:	ror    $0x11,%r8
   0x7ffff6ce02b0 <__longjmp+16>:	xor    %fs:0x30,%r8
   0x7ffff6ce02b9 <__longjmp+25>:	ror    $0x11,%r9
   0x7ffff6ce02bd <__longjmp+29>:	xor    %fs:0x30,%r9
   0x7ffff6ce02c6 <__longjmp+38>:	ror    $0x11,%rdx
   0x7ffff6ce02ca <__longjmp+42>:	xor    %fs:0x30,%rdx
   0x7ffff6ce02d3 <__longjmp+51>:	nop
   0x7ffff6ce02d4 <__longjmp+52>:	mov    (%rdi),%rbx
   0x7ffff6ce02d7 <__longjmp+55>:	mov    0x10(%rdi),%r12
   0x7ffff6ce02db <__longjmp+59>:	mov    0x18(%rdi),%r13
   0x7ffff6ce02df <__longjmp+63>:	mov    0x20(%rdi),%r14
   0x7ffff6ce02e3 <__longjmp+67>:	mov    0x28(%rdi),%r15
   0x7ffff6ce02e7 <__longjmp+71>:	mov    %esi,%eax
   0x7ffff6ce02e9 <__longjmp+73>:	mov    %r8,%rsp
   0x7ffff6ce02ec <__longjmp+76>:	mov    %r9,%rbp
   0x7ffff6ce02ef <__longjmp+79>:	nop
   0x7ffff6ce02f0 <__longjmp+80>:	jmpq   *%rdx
<...>
```

Viva la assembly in libc!

**Comment by glennrp** [comment]
See libpng commit cae11e3e43d9ec7cf0c9550c77c10261f119974a this causes libpng to treat benign errors as warnings instead of errors with longjmp.

**Comment by glennrp** [comment]
Looking back, the run for August 29th shows coverage of the loop in question, so it's a new bug.

**Comment by glennrp** [comment]
See libpng commit 0512c63533d20634bad25f133f3094abbae7cda3 which reverts the fuzzer back to a version that did not exhibit the problem.

**Comment by glennrp** [comment]
As far as I can tell, this problem was due to misplacement of a call to png_read_update_info() ahead of png_set_interlace_handling().  Closing this PR.
